lib/assembly/build.ex

```diff
@@ -0,0 +1,43 @@
+defmodule Assembly.Build do
+  @moduledoc """
+  A genserver for builds. It listens for changes to part counts from inventory
+  and updates it's buildable status.
+  """
+
+  use GenServer
+
+  import Logger
+
+  alias Phoeinx.PubSub
```

Comment:
I'm all for using DynamicSupervisor but I think showhorning in pubsub between these is the wrong approach. We're aiming for simpler code not more complex and difficult to trace.

Let's remove the pubsub from this and focus on how we can implement this with fewest moving parts.

Reply:
Why is using `Phoenix.PubSub` shoehorning? It's just an easier way to communicate between processes. The alternative to using pubsub would be for the builds to pool part availability. The alternative for `GenServer` is just having a function that grabs everything when it's called. The logic itself for seeing if a build can be built is pretty easy (this can be cleaned up even more):

```ex
  def ready_builds() do
    builds = Repo.get_ready_builds()

    components =
      builds
      |> Enum.flat_map(&Map.get(&1, :components, []))
      |> Enum.uniq()

    configurations = Repo.get_configurations(components)

    part_counts = Repo.get_sku_counts()

    {ready_builds, incomplete_builds} =
      Enum.reduce(builds, {[], []}, fn build, {rb, ib} ->
        pickable? = Enum.all?(build.components, fn component ->
          configurations
          |> Enum.filter(&(&1.component_id == component))
          |> Enum.any?(fn c ->
            Map.get(part_counts, to_string(c.sku_id), 0) >= c.quantity
          end)
        end)

        if pickable? do
          {rb ++ [build], ib}
        else
          {rb, ib ++ [build]}
        end
      end)
  end
```

Running this function in production (with direct db access) takes about 1.5 seconds.

Reply:
It's shoehorning because we don't need the complexity of pubsub to accomplish the original proposal for using a DynamicSupervisor.

Polling isn't required, the supervisor know its children's pids and can send messages to them as necessary.